File: firmware/esp32_sensor_node/BiometricManager.cpp
#include "BiometricManager.h"
#include <SPIFFS.h>
#include <ArduinoJson.h>
#include <math.h>

static const char* REGISTRY_PATH = "/bio/users.json";

BiometricManager::BiometricManager() : _userCount(0) {
    memset(_users,         0, sizeof(_users));
    memset(_templateCount, 0, sizeof(_templateCount));
    memset(_templates,     0, sizeof(_templates));
}

bool BiometricManager::begin() {
    if (!SPIFFS.begin(true)) {
        Serial.println("[BIO] SPIFFS mount failed");
        return false;
    }
    _loadAll();
    Serial.printf("[BIO] Ready — %d user(s) enrolled\n", _userCount);
    return true;
}

bool BiometricManager::enroll(const char* userId, const char* name,
                               const IrisCapture& iris) {
    if (!iris.valid) {
        Serial.println("[BIO] enroll() rejected: invalid capture");
        return false;
    }

    int idx = _findUser(userId);
    if (idx < 0) {
        if (_userCount >= BIO_MAX_USERS) {
            Serial.println("[BIO] Max users reached");
            return false;
        }
        idx = _userCount++;
        strlcpy(_users[idx].userId, userId, sizeof(_users[idx].userId));
        strlcpy(_users[idx].name,   name,   sizeof(_users[idx].name));
        _users[idx].active        = true;
        _users[idx].templateCount = 0;
        _templateCount[idx]       = 0;
    }

    // Overwrite oldest template when at capacity (ring-buffer)
    uint8_t tIdx = _templateCount[idx] % BIO_MAX_TEMPLATES;
    memcpy(_templates[idx][tIdx], iris.features, sizeof(float) * IRIS_FEAT_DIM);

    if (!_saveTemplate(idx, tIdx)) {
        Serial.printf("[BIO] Failed to save template %d for '%s'\n", tIdx, userId);
        return false;
    }

    if (_templateCount[idx] < BIO_MAX_TEMPLATES) _templateCount[idx]++;
    _users[idx].templateCount = _templateCount[idx];

    saveRegistry();
    Serial.printf("[BIO] Enrolled '%s' template %d/%d\n",
                  userId, tIdx + 1, _templateCount[idx]);
    return true;
}

MatchResult BiometricManager::match(const IrisCapture& iris, float threshold) const {
    MatchResult best;
    best.matched     = false;
    best.score       = 1e9f;
    best.templateIdx = 0;
    best.userId[0]   = '\0';
    best.userName[0] = '\0';

    if (!iris.valid || _userCount == 0) return best;

    for (uint8_t u = 0; u < _userCount; u++) {
        if (!_users[u].active || _templateCount[u] == 0) continue;
        for (uint8_t t = 0; t < _templateCount[u]; t++) {
            float d = _rms(iris.features, _templates[u][t]);
            if (d < best.score) {
                best.score       = d;
                best.templateIdx = t;
                strlcpy(best.userId,   _users[u].userId, sizeof(best.userId));
                strlcpy(best.userName, _users[u].name,   sizeof(best.userName));
            }
        }
    }

    best.matched = (best.score < threshold);
    Serial.printf("[BIO] Match: %s score=%.4f thresh=%.4f → %s\n",
                  best.userId, best.score, threshold,
                  best.matched ? "PASS" : "FAIL");
    return best;
}

bool BiometricManager::isEnrolled(const char* userId) const {
    return _findUser(userId) >= 0;
}

bool BiometricManager::removeUser(const char* userId) {
    int idx = _findUser(userId);
    if (idx < 0) return false;

    for (uint8_t t = 0; t < BIO_MAX_TEMPLATES; t++) {
        String p = _tPath(idx, t);
        if (SPIFFS.exists(p)) SPIFFS.remove(p);
    }

    _users[idx].active        = false;
    _users[idx].templateCount = 0;
    _templateCount[idx]       = 0;

    saveRegistry();
    Serial.printf("[BIO] Removed user '%s'\n", userId);
    return true;
}

bool BiometricManager::saveRegistry() const {
    StaticJsonDocument<2048> doc;
    JsonArray arr = doc.createNestedArray("users");

    for (uint8_t i = 0; i < _userCount; i++) {
        if (!_users[i].active) continue;
        JsonObject o = arr.createNestedObject();
        o["id"]     = _users[i].userId;
        o["name"]   = _users[i].name;
        o["count"]  = _templateCount[i];
        o["active"] = true;
    }

    File f = SPIFFS.open(REGISTRY_PATH, "w");
    if (!f) return false;
    serializeJson(doc, f);
    f.close();
    return true;
}

// ── Private ──────────────────────────────────────────────────────────────────

bool BiometricManager::_loadAll() {
    if (!SPIFFS.exists(REGISTRY_PATH)) return true;  // no users yet — OK

    File f = SPIFFS.open(REGISTRY_PATH, "r");
    if (!f) return false;

    StaticJsonDocument<2048> doc;
    DeserializationError err = deserializeJson(doc, f);
    f.close();
    if (err) {
        Serial.printf("[BIO] Registry parse error: %s\n", err.c_str());
        return false;
    }

    _userCount = 0;
    for (JsonObject u : doc["users"].as<JsonArray>()) {
        if (_userCount >= BIO_MAX_USERS) break;
        uint8_t idx = _userCount++;
        strlcpy(_users[idx].userId, u["id"]   | "", sizeof(_users[idx].userId));
        strlcpy(_users[idx].name,   u["name"] | "", sizeof(_users[idx].name));
        _users[idx].templateCount = u["count"]  | (uint8_t)0;
        _users[idx].active        = u["active"] | false;
        _templateCount[idx]       = 0;
        for (uint8_t t = 0; t < _users[idx].templateCount; t++) _loadTemplate(idx, t);
    }
    return true;
}

int BiometricManager::_findUser(const char* userId) const {
    for (uint8_t i = 0; i < _userCount; i++)
        if (_users[i].active && strcmp(_users[i].userId, userId) == 0) return i;
    return -1;
}

float BiometricManager::_rms(const float* a, const float* b) const {
    float sum = 0.0f;
    for (uint8_t i = 0; i < IRIS_FEAT_DIM; i++) {
        float d = a[i] - b[i]; sum += d * d;
    }
    return sqrtf(sum / IRIS_FEAT_DIM);
}

bool BiometricManager::_saveTemplate(uint8_t ui, uint8_t ti) const {
    File f = SPIFFS.open(_tPath(ui, ti), "w");
    if (!f) return false;
    f.write((const uint8_t*)_templates[ui][ti], sizeof(float) * IRIS_FEAT_DIM);
    f.close();
    return true;
}

bool BiometricManager::_loadTemplate(uint8_t ui, uint8_t ti) {
    String p = _tPath(ui, ti);
    if (!SPIFFS.exists(p)) return false;
    File f = SPIFFS.open(p, "r");
    if (!f) return false;
    size_t n = f.read((uint8_t*)_templates[ui][ti], sizeof(float) * IRIS_FEAT_DIM);
    f.close();
    if (n == sizeof(float) * IRIS_FEAT_DIM) { _templateCount[ui]++; return true; }
    return false;
}

String BiometricManager::_tPath(uint8_t ui, uint8_t ti) const {
    return String("/bio/") + _users[ui].userId + "_t" + ti + ".bin";
}

File: firmware/rpi_sensor_node/biometric_manager.py
import os
import json
import numpy as np
from dataclasses import dataclass, field
from typing import List, Optional

from iris_camera import IrisCapture, IRIS_FEAT_DIM

BIO_MAX_USERS     = 16
BIO_MAX_TEMPLATES = 5
_BIO_DIR          = os.path.join(os.path.dirname(__file__), "bio")
_REGISTRY_PATH    = os.path.join(_BIO_DIR, "users.json")


@dataclass
class UserRecord:
    user_id: str = ""
    name: str = ""
    template_count: int = 0
    active: bool = False


@dataclass
class MatchResult:
    matched: bool = False
    user_id: str = ""
    user_name: str = ""
    score: float = 1e9
    template_idx: int = 0


class BiometricManager:
    def __init__(self):
        self._users: List[UserRecord] = []
        self._templates: List[List[Optional[np.ndarray]]] = []
        self._template_counts: List[int] = []

    def begin(self) -> bool:
        os.makedirs(_BIO_DIR, exist_ok=True)
        self._load_all()
        print(f"[BIO] Ready — {len(self._users)} user(s) enrolled")
        return True

    def enroll(self, user_id: str, name: str, iris: IrisCapture) -> bool:
        if not iris.valid:
            print("[BIO] enroll() rejected: invalid capture")
            return False

        idx = self._find_user(user_id)
        if idx < 0:
            if len(self._users) >= BIO_MAX_USERS:
                print("[BIO] Max users reached")
                return False
            idx = len(self._users)
            self._users.append(UserRecord(user_id=user_id, name=name, active=True))
            self._templates.append([None] * BIO_MAX_TEMPLATES)
            self._template_counts.append(0)

        # Ring-buffer overwrite when at capacity
        t_idx = self._template_counts[idx] % BIO_MAX_TEMPLATES
        self._templates[idx][t_idx] = iris.features.copy()

        if not self._save_template(idx, t_idx):
            print(f"[BIO] Failed to save template {t_idx} for '{user_id}'")
            return False

        if self._template_counts[idx] < BIO_MAX_TEMPLATES:
            self._template_counts[idx] += 1
        self._users[idx].template_count = self._template_counts[idx]

        self._save_registry()
        print(f"[BIO] Enrolled '{user_id}' template {t_idx + 1}/{self._template_counts[idx]}")
        return True

    def match(self, iris: IrisCapture, threshold: float) -> MatchResult:
        best = MatchResult()
        if not iris.valid or not self._users:
            return best

        for u_idx, user in enumerate(self._users):
            if not user.active or self._template_counts[u_idx] == 0:
                continue
            for t_idx in range(self._template_counts[u_idx]):
                tmpl = self._templates[u_idx][t_idx]
                if tmpl is None:
                    continue
                d = self._rms(iris.features, tmpl)
                if d < best.score:
                    best.score       = d
                    best.template_idx = t_idx
                    best.user_id     = user.user_id
                    best.user_name   = user.name

        best.matched = best.score < threshold
        status = "PASS" if best.matched else "FAIL"
        print(f"[BIO] Match: {best.user_id} score={best.score:.4f} thresh={threshold:.4f} → {status}")
        return best

    def user_count(self) -> int:
        return sum(1 for u in self._users if u.active)

    def is_enrolled(self, user_id: str) -> bool:
        return self._find_user(user_id) >= 0

    def remove_user(self, user_id: str) -> bool:
        idx = self._find_user(user_id)
        if idx < 0:
            return False
        for t_idx in range(BIO_MAX_TEMPLATES):
            p = self._template_path(idx, t_idx)
            if os.path.exists(p):
                os.remove(p)
        self._users[idx].active        = False
        self._template_counts[idx]     = 0
        self._save_registry()
        print(f"[BIO] Removed user '{user_id}'")
        return True

    # ── Private ───────────────────────────────────────────────────────────────

    def _rms(self, a: np.ndarray, b: np.ndarray) -> float:
        """Normalised RMS distance — matches BiometricManager::_rms() exactly."""
        return float(np.sqrt(np.mean((a - b) ** 2)))

    def _find_user(self, user_id: str) -> int:
        for i, u in enumerate(self._users):
            if u.active and u.user_id == user_id:
                return i
        return -1

    def _template_path(self, u_idx: int, t_idx: int) -> str:
        return os.path.join(_BIO_DIR, f"{self._users[u_idx].user_id}_t{t_idx}.npy")

    def _save_template(self, u_idx: int, t_idx: int) -> bool:
        try:
            np.save(self._template_path(u_idx, t_idx), self._templates[u_idx][t_idx])
            return True
        except Exception as e:
            print(f"[BIO] Save error: {e}")
            return False

    def _load_all(self):
        if not os.path.exists(_REGISTRY_PATH):
            return
        try:
            with open(_REGISTRY_PATH, "r") as f:
                data = json.load(f)
        except Exception as e:
            print(f"[BIO] Registry parse error: {e}")
            return

        for u in data.get("users", []):
            if len(self._users) >= BIO_MAX_USERS:
                break
            user = UserRecord(
                user_id=u.get("id", ""),
                name=u.get("name", ""),
                template_count=u.get("count", 0),
                active=u.get("active", False),
            )
            self._users.append(user)
            slots: List[Optional[np.ndarray]] = [None] * BIO_MAX_TEMPLATES
            loaded = 0
            for t_idx in range(user.template_count):
                p = os.path.join(_BIO_DIR, f"{user.user_id}_t{t_idx}.npy")
                if os.path.exists(p):
                    try:
                        slots[t_idx] = np.load(p)
                        loaded += 1
                    except Exception:
                        pass
            self._templates.append(slots)
            self._template_counts.append(loaded)

    def _save_registry(self):
        data = {"users": [
            {"id": u.user_id, "name": u.name,
             "count": u.template_count, "active": u.active}
            for u in self._users
        ]}
        try:
            with open(_REGISTRY_PATH, "w") as f:
                json.dump(data, f, indent=2)
        except Exception as e:
            print(f"[BIO] Registry save error: {e}")
